File: Tennis/first_try/TennisKata/spec/tennis_spec.rb
# To change this template, choose Tools | Templates
# and open the template in the editor.

require 'tennis'

describe Tennis do
  before(:each) do
    @tennis = Tennis.new
  end

  describe "score" do
    {
      {:player_one => 0, :player_two => 0} => "love - love",
      {:player_one => 1, :player_two => 0} => "fifteen - love",
      {:player_one => 1, :player_two => 1} => "fifteen - fifteen",
      {:player_one => 2, :player_two => 1} => "thirty - fifteen",
      {:player_one => 2, :player_two => 2} => "thirty - thirty",
      {:player_one => 3, :player_two => 2} => "forty - thirty",
      {:player_one => 0, :player_two => 1} => "love - fifteen",
      {:player_one => 1, :player_two => 2} => "fifteen - thirty",
      {:player_one => 2, :player_two => 3} => "thirty - forty",
      {:player_one => 2, :player_two => 0} => "thirty - love",
      {:player_one => 0, :player_two => 2} => "love - thirty",
      {:player_one => 3, :player_two => 0} => "forty - love",
      {:player_one => 0, :player_two => 3} => "love - forty",
    }.each do | balls, expected_result |
      it "should return '#{expected_result}' when #{balls[:player_one]} - #{balls[:player_two]}" do
        @tennis.play balls[:player_one], balls[:player_two]
        @tennis.score.should eql expected_result
      end
    end
  end

  describe "deuce" do
    {
      {:player_one => 3, :player_two => 3} => "deuce",
      {:player_one => 4, :player_two => 4} => "deuce",
      {:player_one => 5, :player_two => 5} => "deuce",
      {:player_one => 6, :player_two => 6} => "deuce",
      {:player_one => 7, :player_two => 7} => "deuce",
    }.each do | balls, expected_result |
      it "should return '#{expected_result}' when #{balls[:player_one]} - #{balls[:player_two]}" do
        @tennis.play balls[:player_one], balls[:player_two]
        @tennis.score.should eql expected_result
      end
    end
  end

  describe "advantage" do
    {
      {:player_one => 4, :player_two => 3} => "advantage player one",
      {:player_one => 3, :player_two => 4} => "advantage player two",
      {:player_one => 5, :player_two => 4} => "advantage player one",
      {:player_one => 4, :player_two => 5} => "advantage player two",
      {:player_one => 7, :player_two => 6} => "advantage player one",
      {:player_one => 6, :player_two => 7} => "advantage player two",
      
    }.each do | balls, expected_result |
      it "should return '#{expected_result}' when #{balls[:player_one]} - #{balls[:player_two]}" do
        @tennis.play balls[:player_one], balls[:player_two]
        @tennis.score.should eql expected_result
      end
    end
  end

  describe "win" do
    {
      {:player_one => 4, :player_two => 0} => "player one wins",
      {:player_one => 0, :player_two => 4} => "player two wins",
#      {:player_one => 5, :player_two => 4} => "advantage player one",
#      {:player_one => 4, :player_two => 5} => "advantage player two",
#      {:player_one => 7, :player_two => 6} => "advantage player one",
#      {:player_one => 6, :player_two => 7} => "advantage player two",
    }.each do | balls, expected_result |
      it "should return '#{expected_result}' when #{balls[:player_one]} - #{balls[:player_two]}" do
        @tennis.play balls[:player_one], balls[:player_two]
        @tennis.score.should eql expected_result
      end
    end
  end

end

File: Tennis/javascript/TennisKata/spec/TennisSpec.js
describe("Tennis", function() {
  var tennis;

  beforeEach(function() {
    tennis = makeTennis();
  });

  describe("regular match", function() {
    [
      {playerOne: 0, playerTwo: 0, result: 'love - love' },
      {playerOne: 0, playerTwo: 1, result: 'love - fifteen' },
      {playerOne: 1, playerTwo: 0, result: 'fifteen - love' },
      {playerOne: 1, playerTwo: 1, result: 'fifteen - fifteen' },
      {playerOne: 1, playerTwo: 2, result: 'fifteen - thirty' },
      {playerOne: 2, playerTwo: 1, result: 'thirty - fifteen' },
      {playerOne: 2, playerTwo: 2, result: 'thirty - thirty' },
      {playerOne: 3, playerTwo: 2, result: 'forty - thirty' },
      {playerOne: 2, playerTwo: 3, result: 'thirty - forty' }
    ].forEach(function(testGame) {it ("should return " + testGame.result + " when score is " + testGame.playerOne + " - " + testGame.playerTwo, function() {
        tennis.game(testGame.playerOne, testGame.playerTwo);
        expect(tennis.score()).toBe(testGame.result);
      });
    });
  });

  describe("deuce", function() {
    [
      {playerOne: 3, playerTwo: 3, result: 'deuce'},
      {playerOne: 4, playerTwo: 4, result: 'deuce'},
      {playerOne: 5, playerTwo: 5, result: 'deuce'},
      {playerOne: 6, playerTwo: 6, result: 'deuce'},
      {playerOne: 7, playerTwo: 7, result: 'deuce'},
      {playerOne: 8, playerTwo: 8, result: 'deuce'},
      {playerOne: 9, playerTwo: 9, result: 'deuce'},
      {playerOne: 10, playerTwo: 10, result: 'deuce'}
    ].forEach(function(testGame) {it ("should return " + testGame.result + " when score is " + testGame.playerOne + " - " + testGame.playerTwo, function() {
        tennis.game(testGame.playerOne, testGame.playerTwo);
        expect(tennis.score()).toBe(testGame.result);
      });
    });
  });

  describe("advantage", function() {
    [
      {playerOne: 4, playerTwo: 3, result: 'advantage player one'},
      {playerOne: 3, playerTwo: 4, result: 'advantage player two'},
      {playerOne: 5, playerTwo: 4, result: 'advantage player one'},
      {playerOne: 4, playerTwo: 5, result: 'advantage player two'},
      {playerOne: 6, playerTwo: 5, result: 'advantage player one'},
      {playerOne: 5, playerTwo: 6, result: 'advantage player two'},
      {playerOne: 7, playerTwo: 6, result: 'advantage player one'},
      {playerOne: 6, playerTwo: 7, result: 'advantage player two'},
      {playerOne: 8, playerTwo: 7, result: 'advantage player one'},
      {playerOne: 7, playerTwo: 8, result: 'advantage player two'}
    ].forEach(function(testGame) {it ("should return " + testGame.result + " when score is " + testGame.playerOne + " - " + testGame.playerTwo, function() {
        tennis.game(testGame.playerOne, testGame.playerTwo);
        expect(tennis.score()).toBe(testGame.result);
      });
    });
  });

  describe("winning", function() {
    [
      {playerOne: 4, playerTwo: 0, result: 'game player one'},
      {playerOne: 0, playerTwo: 4, result: 'game player two'}
    ].forEach(function(testGame) {it ("should return " + testGame.result + " when score is " + testGame.playerOne + " - " + testGame.playerTwo, function() {
        tennis.game(testGame.playerOne, testGame.playerTwo);
        expect(tennis.score()).toBe(testGame.result);
      });
    });
  });
});